duracloud/duracloud
# Add S3 Swift storage provider

# What does this Pull Request do?

This commit adds a new storage option - OpenStack Swift with S3
middleware.

We have written a new module called swiftstorageprovider which overrides
the methods in s3storageprovider that Swift does not support.

S3ProviderUtil is now capable of constructing S3 clients compatible with
both AWS and Swift, and handles the options passed in for Swift clients.

Additionally, this commit makes the necessary changes to `pom.xml` so it works with the checkstyle file included in this repo.

# How should this be tested?

To fully test this, one needs an S3 compatible OpenStack Swift Stein backend.

* Create a new account with S3 Swift as one of the storage options.
* Configure the provider with the appropriate access key, secret key, S3 Endpoint and Signer Type.
* The Endpoint URL should be of the form `http(s)://my.s3.endpoint#Region`. The region is optional as it is only necessary for certain signer types. If you don't need to specify a region, omit the `#`.
* Set the same parameters in your `DURACLOUD_TEST_CONFIG` file. Example:
```
{
  "providerCredentials": [
    {
      "type": "SWIFT_S3",
      "credential": {
        "username": "swifts3accesskey",
        "password": "swifts3secretkey"
      }
    },
    {
      "type": "AMAZON_S3",
      "credential": {
        "username": "amazons3accesskey",
        "password": "amazons3secretkey"
      }
    }
  ],
  "swiftEndpoint": "https://your-endpoint#your-region",
  "swiftSignerType": "S3SignerType",
  "queueName": "testing",
  "testEndPoint": {
    "host": "localhost",
    "port": "8080"
  },
  "userCredential": {
    "username": "normaluser",
    "password": "normaluser"
  },
  "adminCredential": {
    "username": "adminuser",
    "password": "adminuser"
  },
  "rootCredential": {
    "username": "rootuser",
    "password": "rootuser"
  }
}
```
* Run integration tests.

# Additional Notes:

This is only compatible with Stein and later releases of Openstack Swift.

Known issues:

* HLS streaming does not work
* Because the management of the audit log space is decoupled from the storage providers, users still need an AWS S3 account for the audit bucket.
* Mill needs some changes before it is compatible.
* The unit tests for `SwiftStorageProvider` will sometimes fail because a second has incremented between the `Date` creation in `testDoSetSpaceProperties()` and `doSetSpaceProperties()`.

Additional documentation is required detailing how to use this storage option.

# Interested parties
@bbranan
